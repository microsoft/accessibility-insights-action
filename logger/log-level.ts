// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export enum LogLevel {
    info,
    warn,
    verbose,
    error,
}
