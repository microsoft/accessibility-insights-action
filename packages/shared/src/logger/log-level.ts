// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export enum LogLevel {
    info,
    warning,
    debug,
    error,
    startGroup,
    endGroup,
}
