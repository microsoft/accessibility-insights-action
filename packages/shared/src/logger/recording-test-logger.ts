// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { LoggerClient } from './logger-client';
import { LogLevel } from './log-level';
import { Logger } from './logger';
import { LoggerProperties } from './logger-properties';
import * as process from 'process';

// We allow this type to be complex in order to produce readable snapshots
// when a test does expect(recordingLogger.recordedLogs()).toMatchSnapshot()
// Particularly, "Simple" logs (messages with no properties) will show up in
// snapshots as simple strings.
export type LogRecord =
    | string
    | {
          message: string;
          properties?: { [name: string]: string };
      }
    | {
          exception: Error;
      };

export class RecordingTestLogger extends Logger {
    private client: RecordingLoggerClient;
    constructor(currentProcess: typeof process = process) {
        const client = new RecordingLoggerClient();
        super([client], currentProcess);
        this.client = client;
        this.initialized = true;
    }

    public recordedLogs(): LogRecord[] {
        return [...this.client.logRecords];
    }
}

class RecordingLoggerClient implements LoggerClient {
    public logRecords: LogRecord[] = [];

    log(
        message: string,
        logLevel: LogLevel,
        properties?: {
            [name: string]: string;
        },
    ): void {
        const messageWithLevel = `[${LogLevel[logLevel]}] ${message}`;
        if (properties) {
            this.logRecords.push({
                message: messageWithLevel,
                properties,
            });
        } else {
            this.logRecords.push(messageWithLevel);
        }
    }
    trackException(error: Error): void {
        this.logRecords.push({
            exception: error,
        });
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/require-await
    async setup(baseProperties?: { [index: string]: string }): Promise<void> {
        throw new Error('Not supported');
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setCustomProperties(properties: LoggerProperties): void {
        throw new Error('Not supported');
    }
}
