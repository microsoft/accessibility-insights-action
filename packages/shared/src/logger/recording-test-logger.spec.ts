// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { LogLevel } from './log-level';
import { RecordingTestLogger } from './recording-test-logger';

describe(RecordingTestLogger, () => {
    it('records different log calls in order with readable snapshotting', () => {
        const testSubject = new RecordingTestLogger();
        testSubject.log('message 1 (info)', LogLevel.info);
        testSubject.logDebug('message 2 (debug)');
        testSubject.logInfo('message 3 (info)');
        testSubject.logInfo('message 4 (info w/props)', { key: 'val' });
        testSubject.trackException(new Error('message 5 (exception)'));
        testSubject.trackExceptionAny('error', 'message 6 (exceptionAny)');
        testSubject.logWarning('message 7 (warning)');
        testSubject.logError('message 8 (error)');
        testSubject.logStartGroup('message 9 (start group)');
        testSubject.logEndGroup();

        expect(testSubject.recordedLogs()).toMatchSnapshot();
    });
});
