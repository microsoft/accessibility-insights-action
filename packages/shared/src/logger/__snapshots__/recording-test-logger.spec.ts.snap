// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordingTestLogger records different log calls in order with readable snapshotting 1`] = `
[
  "[info] message 1 (info)",
  "[debug] message 2 (debug)",
  "[info] message 3 (info)",
  {
    "message": "[info] message 4 (info w/props)",
    "properties": {
      "key": "val",
    },
  },
  {
    "exception": [Error: message 5 (exception)],
  },
  {
    "exception": [ErrorWithCause: message 6 (exceptionAny)],
  },
  "[warning] message 7 (warning)",
  "[error] message 8 (error)",
  "[group] message 9 (start group)",
  "[endgroup] ",
]
`;
